usr/apex-env.8
.TH apex-env 8 27-January-2007 "Debian GNU/Linux"
.SH NAME
apex-env \- user-mode access to APEX flash environment
.SH SYNOPSIS
.B apex-env
COMMAND
[
.B ARGS ...
]

.SH DESCRIPTION

\fBapex-env\fR is a utility for viewing and modifying the environment
used by the APEX Boot Loader to define how the system will boot.  APEX
has a predefined set of environment variables, defined by the
build-time configuration, that control operation of the boot loader.  

Embedded in the APEX image is a linking structure that \fBapex-env\fR
uses to identify the boot loader image and to locate information about
the environment variables as well as the flash region that holds the
user-modified values.  If APEX is built without enabling this
user-modifiable flash region, \fBapex-env\fR will only allow the user
to browse the variables and the default values.

\fBapex-env\fR depends on the \fI/dev/mtd\fR driver to locate the
APEX.  It first looks for a partition named 'Loader', but it will also
check the first partition in case APEX is used as the primary boot
loader.

.SH COMMANDS

.TP
.I "describe [KEY]"
Displays a description of KEY or descriptions of every environment
variable is KEY is omitted.

.TP
.I "dump"
Displays the contents of the flash environment region in a
hexadecimal/ASCII format useful for troubleshooting.

.TP
.I "eraseenv"
Clear the environment region by setting all of the bytes to 0xff.
Use this command with care.  Assuming that APEX was properly built for
your system, it still relies on the caching feature of the MTD block
driver to modify only a portion of a flash eraseblock.  Be mindful
that interrupting this command after the block is erased and before
the important data has been restored could render the system
unbootable.

.TP
.I "printenv [KEY]"
Displays the value of KEY or the values of every environment variable
if KEY is omitted.  The separator between the key and the value is =*
for default values or just = when the default value has been
overridden by a value saved in the flash environment.

.TP
.I "setenv KEY VALUE"
Overrides the default value of KEY with VALUE.  Use quoting to prevent
the shell from splitting VALUE into separate words and from
interpolation.  

.TP
.I "unsetenv KEY"
Restores the default value of KEY by marking any value stored in flash
as deleted.

.SH EXAMPLES

.TP
apex-env printenv
Show all of the environment variables and the current values for each.

.TP
apex-env setenv startup 'copy $kernelsrc $bootaddr ; boot'
Change the startup variable to copy the kernel and immediately boot.

.TP
apex-env dump | head
Examine the first few bytes of the raw environment data.

.SH AUTHOR
Marc Singer <elf@debian.org>



ChangeLog
2007-02-02  Marc Singer  <elf@zealous.synapsedev.com>

	* src/apex/cmd-wait.c: Allowing a zero TIMEOUT. 

	* src/apex/command.c (TIMECMD_INIT): Added command execution timer.

2007-01-31  Marc Singer  <elf@cerise.buici.com>

	* debian/rules: Cross-building support.

	* src/mach-ixp42x/serial.c: Minor change to the ixp42x serial
	driver to raise RTS for the same of systems that use it.

2007-01-30  Marc Singer  <elf@buici.com>

	* src/arch-arm/lib/lib1funcs.S: New version of lib1funcs so that
	armel is happy.  This code was pulled directly from the Linux
	kernel source.

2007-01-29  Marc Singer  <elf@cerise.buici.com>

	* Added alternatives to Debian configs.

2007-01-29  Marc Singer  <elf@zealous.synapsedev.com>

	* Makefile (SUBLEVEL): Rolled rev for block cache fix.

	* src/drivers/drv-ext2.c: inode/block cache flush was broken.
	Now, it is not.

2007-01-27  Marc Singer  <elf@buici.com>

	* usr/apex-env.8: Man page.

	* usr/link.h: Added cleanup code to Link object.  It isn't
	strictly necessary, but it is good hygiene.

2007-01-27  Marc Singer  <elf@cerise.buici.com>

	* Makefile: 1.4.12 for new environment program and to prepare for
	publishing a new release.

	* usr/apex-env.cc: New user-mode program for updating the APEX
	environment once the system is booted.  It replaces the printenv
	program that was only able to print the environment as it stood,
	combining the flash data with the defaults from APEX.  This
	program is smart enough to be able to set/unset variables and
	erase the whole environment.

2007-01-25  Marc Singer  <elf@zealous.synapsedev.com>

	* src/mach-lh7a40x/clcdc-companion.c: Removed the DDS hack form
	the LCD init on Companion.  No DDS circuit present anymore.

2007-01-23  Marc Singer  <elf@cerise.buici.com>

	* src/lib/memtest.c (memory_test_0): Revised the memory test
	walking bit loop on advice from a collegue.  Made the full memory
	test an option on top of the basic memory test.

2007-01-23  Marc Singer  <elf@zealous.synapsedev.com>

	* src/mach-lh7a40x/variation-companion.c (variation_init): Added
	variation support to the Companion.  Pressing three keys at
	power-on will activate the variation boot.

2007-01-13  Marc Singer  <elf@cerise.buici.com>

	* src/arch-arm/entry/reset.c (__section): Slight change to MMU
	disabling code.

	* src/arch-arm/entry/mmu-xscale.h (CACHE_CLEAN_D): Revised cache
	clean code to match Nicolas Pitre's work in the Linux kernel.

2007-01-09  Marc Singer  <elf@zealous.synapsedev.com>

	* src/drivers/drv-fis.c (end_of_table): Added back the
	end-of-table check.  We're now checking the same way that the
	kernel checks.
	
2007-01-09  Marc Singer  <elf@cerise.buici.com>

	* Makefile (SUBLEVEL): 1.4.11, to fix partition table swapped-ness
	check.

	* src/drivers/drv-fis.c (prescan_directory): Fixed the partition
	table swapped-ness check to a more robust comparison of the
	partition table entry's erase block with the erase block from
	which the table is being read.

2007-01-08  Marc Singer  <elf@cerise.buici.com>

	* 1.4.10.
		
	* src/drivers/drv-fis.c (deleted_entry): Replaced the
	end_of_table() checks with deleted_entry checks.  Really, we
	cannot be sure that we've seen the end of the table since a 0xff
	in the first byte indicates a deleted entry and we have to check
	for that first.  We previously thought that (u32)~0 was the end of
	the table, but we cannot now be sure.  Argh.

	* Applyied nas100d and dsmg600 patches as well as a slew of config
	file patches courtesy of rwhitby.

	* src/drivers/drv-nor-cfi.c (SWAP_ONE): Fixed the nor flash driver
	to properly write when in an endian mismatched mode.  This may be
	peculiar to the nslu2 (and other ixp4xx systems) where flash is
	assumed to be written in one endian-ness while the system is
	executing in the opposite endian-ness.
  
	* src/mach-ixp42x/variation-nslu2.c (variation_init): Variation
	now uses the fact that the power button is depressed (instead of
	the reset button).

	* src/lib/dump.c (dumpw): Fixed the printability check on dump so
	that we don't write non-ascii characters.

	* src/lib/env.c (env_check_magic): Streamlined the environment
	magic number check.

2007-01-06  Marc Singer  <elf@cerise.buici.com>

	* src/apex/env.c: Many more environment variables may be defined.

	* src/apex/cmd-boot.c (atag_commandline): boot command performs
	lookup of the default command line that allows for alternatives.

	* src/arch-arm/Kconfig: Reworking of the environment configuration
	section.  We now have options to set alternative kernel command
	line, kernel source region, and ramdisk region.  Default variation
	suffix is a config parameter.  The default startup command is more
	clearly controlled.

	* src/mach-ixp42x/variation-nslu2.c (variation_init): Added
	variation for nslu2.  Holding the reset button while booting will
	select the alternative boot and set the variation to the
	configured value.

	* src/apex/command.c (expand_variables): Added '-' to the list of
	acceptable characters when scanning a command for variables to
	expand.  This is helpful for variations.

	* src/lib/lookup.c (lookup_alias_or_env): Variation support added
	to the standard lookup function.

2006-12-13  Marc Singer  <elf@zealous.synapsedev.com>

	* src/apex/command.c (parse_command): Added '-' command prefix
	that tells APEX to ignore the return value of a command.  This is
	helpful in the startup command. 

2006-11-07  Marc Singer  <elf@zealous.synapse.com>

	* Makefile (SUBLEVEL): Rolling rev for memory test.

	* src/lib/memtest.c (memory_test_0): memory test for system setup.
	It checks SDRAM before copying the loader post-bootstrap.  Works
	generally on any target, but requires that the
	initialize_bootstrap () function return TRUE when it initializes
	SDRAM.
  
2006-10-25  Marc Singer  <elf@zealous.synapse.com>

	* src/drivers/drv-nand.c (cmd_nand): NAND commands moved to a
	configurable option. 
	(nand_report): Fixed the display of the NAND status.


usr/apex-env.cc
/* apex-env.cc

   written by Marc Singer
   17 Jan 2007

   Copyright (C) 2007 Marc Singer

   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License as
   published by the Free Software Foundation; either version 2 of the
   License, or (at your option) any later version.

   This program is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
   USA.

   -----------
   DESCRIPTION
   -----------

   This program grants access to the APEX environment from user-mode.
   Most of the interest part of the code are in the Link class that
   handles the IO with flash and understands the various protocols and
   formats.  This is merely a driver that calls the Link methods.

   TODO
   ----

   o Force option should be added to allow eraseenv to erase when
     there is non-APEX data in the region.  This doesn't impeded
     normal use of this application.

*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <argp.h>

#include "link.h"

//#define TALK

#if defined (TALK)
# define PRINTF(f ...)	printf(f)
#else
# define PRINTF(f ...)
#endif


const char* argp_program_version = "apex-env 1.0";

struct arguments {
  arguments () {
    bzero (this, sizeof (*this)); }

  int argc;
  const char* argv[3];		// command, key, value
};

struct command {
  const char* sz;
  void (*func) (Link&, int, const char**);
};

static error_t arg_parser (int key, char* arg, struct argp_state* state)
{
  struct arguments& args = *(struct arguments*) state->input;

  switch (key) {
  case ARGP_KEY_ARG:
    if (args.argc >= sizeof (args.argv)/sizeof (*args.argv))
      argp_usage (state);
    args.argv[args.argc++] = arg;
    break;

  case ARGP_KEY_END:
    // *** FIXME: check for the right number of arguments per command?
    break;

  default:
    return ARGP_ERR_UNKNOWN;
  }
  return 0;
}

static struct argp argp = {
  0, arg_parser,
  "COMMAND [ARG ...]",
  "apex-env provides for user-mode access to APEX boot loader environment\n"
  "\n"
  "Commands:\n"
  "  describe [KEY]\t- describe KEY or all variables\n"
  "  dump\t\t\t- hexadecimal/ascii dump of environment region\n"
  "  printenv [KEY]\t- print KEY or all variables\n"
  "  setenv KEY VALUE\t- set variable KEY to VALUE\n"
  "  eraseenv\t\t- erase environment region\n"
};

void cmd_printenv (Link& link, int argc, const char** argv)
{
  switch (argc) {
  case 1:
    link.show_environment ();
    break;
  case 2:
    link.printenv (argv[1]);
    break;
  default:
    throw "incorect number of command arguments";
  }
}

void cmd_describe (Link& link, int argc, const char** argv)
{
  if (argc > 2)
    throw "incorrect number of command arguments";

  link.describe ((argc > 1) ? argv[1] : NULL);
}

void cmd_dump (Link& link, int argc, const char** argv)
{
  if (argc == 1)
    link.dump ();
  else
    return throw "incorrect number of command arguments";
}

void cmd_setenv (Link& link, int argc, const char** argv)
{
  if (argc == 3)
    link.setenv (argv[1], argv[2]);
  else
    return throw "incorrect number of command arguments";
}

void cmd_unsetenv (Link& link, int argc, const char** argv)
{
  if (argc == 2)
    link.unsetenv (argv[1]);
  else
    throw "incorrect number of command arguments";
}

void cmd_eraseenv (Link& link, int argc, const char** argv)
{
  if (argc == 1)
    link.eraseenv ();
  else
    throw "incorrect number of command arguments";
}

static struct command commands[] = {
  { "describe",		cmd_describe },
  { "dump",		cmd_dump },
  { "printenv",		cmd_printenv },
  { "setenv",		cmd_setenv },
  { "unsetenv",		cmd_unsetenv },
  { "eraseenv",		cmd_eraseenv },
};

int main (int argc, char** argv)
{
  struct arguments args;
  argp_parse (&argp, argc, argv, 0, 0, &args);

  try {
    Link link;

    link.open ();

    if (args.argc == 0) {
      ++args.argc;
      args.argv[0] = "printenv"; // Default command
    }

    for (int i = 0; i < sizeof (commands)/sizeof (*commands); ++i)
      if (strcasecmp (args.argv[0], commands[i].sz) == 0) {
	commands[i].func (link, args.argc, args.argv);
	return 0;
      }
    throw "unknown command";
  }
  catch (char const* sz) {
    printf ("error: %s\n", sz);
    argp_help (&argp, stderr, ARGP_HELP_SEE, "apex-env");
  }

  return 0;
}
